packages/react/src/components/popover/useObserver.ts

```diff
@@ -0,0 +1,31 @@
+import { RefObject, useEffect, useRef } from 'react';
+
+/**
+ * Observes `targets` for intersection changes.
+ * @param targets `RefObject` of element to observe.
+ * @param onChange Called when the threshold changes. Either the element just became visible or the opposite.
+ * @param debounceTime Time in milliseconds to debounce events.
+ */
+export function useObserver<T extends Element>(
+  onChange: (entry: IntersectionObserverEntry) => void,
+  targets: RefObject<T>[],
+  debounceTime?: number
+) {
+  const timeout = useRef<number>();
```

Comment:
Why are we using a ref here for the timeout rather than some state?

Reply:
If we used state, every time the `effect` was called we'd update it, triggering a re-run of the entire hook.
Furthermore, that state would need to be part of the `effect`'s dependencies, so we'd have to destroy the previous Observer and create a new one.

Using `ref` makes it so we prevent a lot of render cycles, while still having access to the latest timeout "id" which is all we want.